mne/preprocessing/hfc.py

```diff
@@ -0,0 +1,135 @@
+# Authors: George O'Neill <g.o'neill@ucl.ac.uk>
+#
+# License: BSD-3-Clause
+
+import numpy as np
+from .maxwell import _prep_mf_coils, _sss_basis
+from ..forward import _prep_meg_channels
+from ..io.pick import pick_types
+from ..io.proj import Projection
+from ..utils import verbose
+
+
+@verbose
+def compute_proj_hfc(info, order=1, picks=None, exclude_bads=True,
+                     accuracy='accurate', verbose=None):
+    """Generate projectors to perform homogeneous/harmonic correction to data.
+
+    Remove evironmental fields from magentometer data by assuming it is
+    explained as a homogeneous :footcite:`TierneyEtAl2021` or harmonic field
+    :footcite:`TierneyEtAl2022`. Useful for arrays of OPMs.
+
+    Parameters
+    ----------
+    info : instance of Info
+        Info from an instance of Raw (preferable).
+    order : int
+        The order of the spherical harmonic basis set to use. Set to 1 to use
+        only the homogeneous field component (default), 2 to add gradients, 3
+        to add quadrature terms etc.
+    picks : int | array-like
+        List of channels indices to compute projeciton on. Default behaviour
+        is to select meg (non-reference) with known positions.
+    exclude_bads : bool
+        Do not include bad channels in the projection, (default: True).
+    accuracy : str
+        Can be ``"point"``, ``"normal"`` or ``"accurate"`` (default), defines
+        which level of coil definition accuracy is used to generate model.
+    %(verbose)s
+
+    Returns
+    -------
+    projs: list
+        List of projection vectors.
+
+    Notes
+    -----
+    To apply the projectors to a dataset, use
+    ``[DATA].add_proj(projs).apply_proj()``.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    if picks is None:
+        if exclude_bads is True:
+            picks = pick_types(info, meg=True)
+        else:
+            picks = pick_types(info, meg=True, exclude=None)
+    picks = _filter_channels_with_positions(info, picks)
```

Comment:
Rather than this `_filter_channels_with_positions` I'd rather just check that all `loc` are finite and raise an error if not. It's safer. If I pass `picks=[0, 1, 2]` and my first channel has a bad `loc`, the current code will silently allow this. It's better to raise an error.

And for the default `picks='meg'` all channels should have positions for any sane data (except maybe with ref sensors, in case we should add `ignore_ref=True` kwarg by default, which you can also pass to `_picks_to_idx` IIRC).

Reply:
I've just been having a a play wth _picks_to_idx. I can select non-ref (`picks='meg'`), ref (`picks='ref_meg'`) or every channel including non-MEG (`picks='ref_meg'`). Is there a ways to just get meg and ref_meg if both are needed? There is a `with_ref_meg` kwarg but that appears to do nothing? Is this a case of needing two calls if for some reason we want to apply the correction to reference channels?

Reply:
>all channels should have positions for any sane data (except maybe with ref sensors).

Our data isn't particularly sane. All ref channels in our system do not have positions but as our setup in london is pretty dynamic not all sensors without positions will have been laelled as reference sensors ;)

Reply:
Still, rather than make this one new preprocessinv function handle this oddity this particular way (which will break a lot of other stuff downstream loudly or, more dangerously, silently) let's raise an error. We should separately give you a way to fix your data so that it works with all functions including this one correctly. For example a preproc function to take any MEG channel with an invalid position and label it as a reference sensor or so might make sense.

Reply:
>  Is there a ways to just get meg and ref_meg if both are needed

I think `('meg', 'ref_meg')` maybe? If this doesn't work I can look more on Monday